[Driver] Always link compiler_rt on Darwin

jrose-apple opened the pull request:
Turns out it's needed for normal builtins that can appear in inlinable functions, including Objective-C's `@available`. Clang always links it unconditionally, so so should Swift.

~Note that this does mean you have to build compiler_rt to get a successful test run on Apple platforms. That was always true if you wanted the sanitizer tests to work, though.~

rdar://problem/41911599

jrose-apple committed:
[Driver] Always link compiler_rt on Darwin

Turns out it's needed for normal builtins that can appear in inlinable
functions, including Objective-C's @available. Clang always links it
unconditionally, so so should Swift.

Note that this does mean you have to build compiler_rt to get a
successful test run on Apple platforms. That was always true if you
wanted the sanitizer tests to work, though.

rdar://problem/41911599

jrose-apple commented:
@swift-ci Please test

jrose-apple committed:
[Driver] Only link to compiler_rt if present for the target platform

Tweak the tests to check this correctly.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
apple/swift-source-compat-suite#213
@swift-ci Please test source compatibility

davidungar commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -84,22 +84,29 @@ toolchains::Darwin::constructInvocation(const InterpretJobAction &job,
+      return "iossim";
+    LLVM_FALLTHROUGH;
   case DarwinPlatformKind::IPhoneOS:
     return "ios";
```
What you *really* want here IMO is Swift's switch on tuple, so you could switch on both.
Failing that, it would be nice to make the symmetry wrt `distinguishSimulator` more apparent.
E.g.: 
```
DarwinPlatformKind mapSimulatorToOS(DarwinPlatformKind kind) {
switch kind {
  case DarwinPlatformKind::MacOS: return DarwinPlatformKind::MacOS
  case IPhoneOSSimulator: return IPhoneOS
  case IPhoneOS: return IPhoneOS
...
}

static StringRef
getDarwinLibraryNameSuffixForTriple(const llvm::Triple &triple,
                                    bool distinguishSimulator = true) {
  const DarwinPlatformKind kind = getDarwinPlatformKind(triple)
  const DarwinPlatformKind effectiveKind = distinguishSimulator ? kind : mapSimulatorToOS(kind)
  switch (effectiveKind) {
  case DarwinPlatformKind::MacOS:
    return "osx";
  case DarwinPlatformKind::IPhoneOSSimulator:
    return "iossim";
  case DarwinPlatformKind::IPhoneOS:
    return "ios";
   ...
  }
````
I like this because IMO it reduces the possibility for future errors.

davidungar commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -262,7 +269,22 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+        ".a");
+  if (llvm::sys::fs::exists(CompilerRTPath))
+    Arguments.push_back(context.Args.MakeArgString(CompilerRTPath));
+
```
The routine this is in, `constructInvocation` is already too long for me to understand without breaking it down. At the least, it would be great to factor this paragraph out into a separate function. At best, in order to ease the way for others to work on this code, it would help a lot to break this `constructInvocation` function down into smaller pieces.

davidungar replied:
Ahh, yes, I understand about cherry-picking. Sigh. Will check the other box.

davidungar commented on test/Driver/linker-clang_rt.swift:
```diff
@@ -0,0 +1,36 @@
+// REQUIRES: OS=macosx
```
What is this test testing? What is essential vs incidental? Comments to that effect would be helpful.

davidungar commented on test/lit.cfg:
```diff
@@ -945,8 +945,9 @@ def source_compiler_rt_libs(path):
 
-source_compiler_rt_libs(make_path(test_resource_dir, 'clang', 'lib',
-                        platform.system().lower()))
+compiler_rt_dir = make_path(test_resource_dir, 'clang', 'lib',
```
Would it be possible to add a comment to explain the why here?

davidungar replied:
LOL! I need to add a comment to explain my request for a comment!
My thought was to add comments to this file to explain what is going on w.r.t. the new feature being added. (I know nothing about lit.cfg files and thought maybe there were breadcrumbs you could add.)

jrose-apple replied:
This feels like another case where there's not a reasonable comment. This part didn't change, and the part below builds a path and tests if it exists.

davidungar commented on validation-test/Driver/Inputs/clang_rt-helper.h:
```diff
@@ -0,0 +1,9 @@
+#include <stdbool.h>
+
+static inline bool testNewOS() {
```
Is this testing the OS, or testing the availability of the OS? If the latter, the current name could be improved.

davidungar commented on validation-test/Driver/clang_rt.swift:
```diff
@@ -0,0 +1,11 @@
+
+// Just make sure we can build and link successfully.
+
+if testNewOS() {
```
Or maybe the name should be `isTestingNewOS`?

davidungar reviewed (changes requested):
I know this isn't a huge change, but that constructInvocation function really needs to be broken up. All my other comments are optional. I realize this change is merely adding to the long function, but I would still feel so much better if you could take this opportunity to improve it.

davidungar reviewed (commented):
Too bad this will be cherry-picked and so too risky to bust up constructInvocation. Thanks for adding the other clarifications (i.e. comments, and refactoring the switch).

jrose-apple committed:
Add swift::getNonSimulatorPlatform for DarwinPlatformKind

And use it in getDarwinLibraryNameSuffixForTriple, making the logic
there clearer.

Suggested by David U!

jrose-apple committed:
[test] Tidy up newly-added clang_rt Driver tests to be a bit clearer

Per feedback from David U.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility
